mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+        """Removes an agent from the cell."""
+        self.agents.remove(agent)
+        agent.cell = None
+
+    @property
+    def is_empty(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
+    def neighborhood(self, radius=1, include_center=False):
+        if radius == 0:
+            return {self: self.agents}
+        if radius == 1:
+            return CellCollection(
+                {neighbor: neighbor.agents for neighbor in self._connections}
+            )
+        else:
+            neighborhood = {}
+            for neighbor in self._connections:
+                neighborhood.update(neighbor.neighorhood(radius - 1, include_center))
+            if not include_center:
+                neighborhood.pop(self, None)
+            return CellCollection(neighborhood)
+
+
+class CellCollection:
+    def __init__(self, cells: dict[Cell, list[Agent]] | Iterable[Cell]) -> None:
+        if isinstance(cells, dict):
+            self._cells = cells
+        else:
+            self._cells = {cell: cell.agents for cell in cells}
+
+    def __iter__(self):
+        return iter(self._cells)
+
+    def __getitem__(self, key):
+        return self._cells[key]
+
+    @cached_property
+    def __len__(self):
+        return len(self._cells)
+
+    def __repr__(self):
+        return f"CellCollection({self._cells})"
+
+    @cached_property
+    def cells(self):
+        return list(self._cells.keys())
+
+    @property
+    def agents(self):
+        return itertools.chain.from_iterable(self._cells.values())
+
+    def select_random_cell(self):
+        return random.choice(self.cells)
+
+    def select_random_agent(self):
+        return random.choice(list(self.agents))
+
+    def select(self, filter_func: Optional[Callable[[Cell], bool]] = None, n=0):
+        if filter_func is None and n == 0:
+            return self
+
+        return CellCollection(
+            {
+                cell: agents
+                for cell, agents in self._cells.items()
+                if filter_func is None or filter_func(cell)
+            }
+        )
+
+
+class Space:
+    cells: dict[Coordinate, Cell] = {}
+
+    def _connect_single_cell(self, cell):
+        ...
+
+    @cached_property
+    def all_cells(self):
+        return CellCollection({cell: cell.agents for cell in self.cells.values()})
+
+    def __iter__(self):
+        return iter(self.cells.values())
+
+    def __getitem__(self, key):
+        return self.cells[key]
+
+    def move_agent(self, agent: Agent, pos: Coordinate) -> None:
+        """Move an agent from its current position to a new position."""
+        self.cells[pos].add_agent(agent)
```

Comment:
This is incomplete right?

Reply:
We debated some of this in #1953. I see several solutions

1. Use cell properties on the agent and handle the remove from other cell within the `cell.setter` property. So

```python
# within an agent 
self.cell = self.cell.neighborhood(radius=1).select_random()
```

I like this code. It is clean, expressive and concise. A different version would be to have a `move` method on the agent Either version however breaks fundamentally how MESA has been handling movement in spaces. 

2. Have a separate `move_agent_to(agent:Agent, cell:Cell)` function. This handles the removing from cell and adding to new cell. This is less breaking. It is in line with NetLogo (a reason in my view to not do this :grin:). And is not as potentially tying movement to cells or grids where it does not conceptually belong. 

3. Do it in spaces, as currently done.
4. handle movement within add_agent. Again implicit so bad idea.